File: frontend/src/app/domain/use-cases/get-products.use-case.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Product } from '../entities/product.entity';
import { Page } from '../models/page.model';
import { ProductRepository } from '../repositories/product.repository';

@Injectable()
export class GetProductsUseCase {
  constructor(private productRepository: ProductRepository) {}

  execute(page: number = 0, size: number = 20, sort?: string): Observable<Page<Product>> {
    return this.productRepository.findAll(page, size, sort);
  }

  executeByCategory(category: string, page: number = 0, size: number = 20): Observable<Page<Product>> {
    return this.productRepository.findByCategory(category, page, size);
  }

  executeByBrand(brand: string, page: number = 0, size: number = 20): Observable<Page<Product>> {
    return this.productRepository.findByBrand(brand, page, size);
  }

  executeSearch(query: string, page: number = 0, size: number = 20): Observable<Page<Product>> {
    return this.productRepository.search(query, page, size);
  }
}

File: backend/src/main/java/com/sportsstore/application/usecase/GetProductsUseCase.java
package com.sportsstore.application.usecase;

import com.sportsstore.application.port.out.ProductRepositoryPort;
import com.sportsstore.domain.entity.Product;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class GetProductsUseCase {

    private final ProductRepositoryPort productRepository;

    public GetProductsUseCase(ProductRepositoryPort productRepository) {
        this.productRepository = productRepository;
    }

    public Page<Product> execute(Pageable pageable) {
        return productRepository.findAll(pageable);
    }

    public Page<Product> executeByCategory(String category, Pageable pageable) {
        return productRepository.findByCategory(category, pageable);
    }

    public Page<Product> executeByBrand(String brand, Pageable pageable) {
        return productRepository.findByBrand(brand, pageable);
    }

    public Page<Product> executeSearch(String query, Pageable pageable) {
        return productRepository.search(query, pageable);
    }
}